### DISCUSSION TITLE
How to increase the data volume more effectively for ASR?

### DISCUSSION THREAD

**Discussion by ahkarami** [NVIDIA-NeMo/NeMo#3657] [Q&A] [answered]
Hi, 
Suppose we have a training data set with 500 hours of audio samples with its associated transcriptions. Now, we want to increase the data volume for increasing the accuracy of our ASR model (_e.g._, our ASR model is a token based citrinet512 or ContextNet and we want to reduce its WER by adding more training data).
My question is that, what is the effective way to increase the size of data? 
I mean that, for example we have 100k hours raw audio in a specific domain, and we want to add _e.g._, just 500 hours more data.  
It is obviously clear that we can add _e.g._, 500 hours more data to it, by randomly select 500 hours from the rest of the 100k hours data and transcribe the new 500 hours data and add them to our data set.
But, is there any more effective way to add data? (_i.e._, selecting intelligently some audio files that these utterances reduces more WER of our model in comparison of just randomly sampled the data).
I have founded that there are also some papers in this domain (not exactly the same issue as I have raised, but similar to it) like the below paper:
[1] **SynthASR: Unlocking Synthetic Data for Speech Recognition**, Amazon Alexa Speech, 2021, [link](https://arxiv.org/pdf/2106.07803.pdf).
Another idea that I have, is that, one can analysis the errors of the model (_e.g._, via the [Speech Data Explorer Tool](https://github.com/NVIDIA/NeMo/tree/main/tools/speech_data_explorer) of the Nemo) and add audio samples that have words in which the model has more error on them. 
Thanks for letting me know if anyone has an idea or method for this issue.
Best

**Comment by titu1994** [comment, marked as answer]
If you have transcribed data in that 100k hours, then the simplest way is perform offline inference with a given model and select sunset of audio with high cer / wer. 

Now if you don't have transcribed 100k hours, you are left with manual approach or perform active learning style techniques. 

Manual approach is as you have described - transcribe all of the audio, use SDE to see where transcript length is way off from audio length (meaning transcript was way off). This may work or not. 

Synthasr is to generate synthetic audio- you've already got 100k hours. 

I am not aware of anything other than SSL that can help when you don't have transcribed audio to automate the comparison and select a subset of the dataset for active learning.

**Reply by ahkarami** [reply]
@titu1994 
Thanks a lot for your answer. 
I have just one another question. Is [your SSL](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/speech_pretraining/speech_pre_training.py) (Nemo SSL) based on a specific paper?
Best

**Reply by ahkarami** [reply]
@titu1994 
excuse me. Do you suggest any paper or repo for active learning in ASR?
Best

**Reply by titu1994** [reply]
I don't know of work on active learning for ASR, but I'm sure it's out there. For SSL, it's an amalgamation of various papers I believe. I'll let @sam1373 explain in detail. Also Sam, we really need a section in the ASR docs for SSL (for users and for our own sake)

**Reply by sam1373** [reply]
Currently SSL for ASR in NeMo is done using contrastive loss in a similar way to [wav2vec 2.0](https://arxiv.org/pdf/2006.11477.pdf), which allows pre-training asr models directly on unlabeled data. You can check out this tutorial for a basic example of pre-training and then fine-tuning a Citrinet model: [link](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Self_Supervised_Pre_Training.ipynb) . We plan to update with more recent advances in ssl soon.

**Reply by ahkarami** [reply]
@sam1373 
sorry, as the follow up of this discussion, I have just one another question. Is there any technical report in Nemo that demonstrates how the SSL pre-training can help the WER of the ASR models?

**Reply by sam1373** [reply]
@ahkarami There isn't a technical report yet, but to give an example, if you pre-train Conformer on full LibriSpeech (~1k hours), you can then fine-tune on just 10 labeled hours and get <20% WER on LS test-other (with no LM). If you train on just the 10 hours, you will have a hard time even reaching 50% WER with no LM. The paper I linked above also has some tables with results of pre-training on full LS and then fine-tuning on different subsets using wav2vec architecture. You can get results that are similar or better (especially with transducer decoder) than that paper if you use Conformer in our repository, and probably slightly worse if you use Citrinet.

**Reply by ahkarami** [reply]
Thanks a lot @sam1373 
Your information was really helpful and valuable. 
Best
